File: schemas/digital-asset-management/mongodb/metadata_schemas.js
// metadata_schemas: Custom metadata field definitions per workspace.
// See README.md for full design rationale.

const mongoose = require("mongoose");

const metadataSchemaSchema = new mongoose.Schema(
  {
    workspace_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Workspace",
      required: true,
    },
    field_name: { type: String, required: true },
    field_label: { type: String, required: true },
    field_type: {
      type: String,
      required: true,
      enum: ["text", "number", "date", "boolean", "single_select", "multi_select"],
    },
    options: { type: mongoose.Schema.Types.Mixed, default: null },
    is_required: { type: Boolean, required: true, default: false },
    display_order: { type: Number, required: true, default: 0 },
  },
  {
    timestamps: { createdAt: "created_at", updatedAt: "updated_at" },
  }
);

// Indexes
metadataSchemaSchema.index(
  { workspace_id: 1, field_name: 1 },
  { unique: true }
);

module.exports = mongoose.model("MetadataSchema", metadataSchemaSchema);

File: schemas/digital-asset-management/spacetimedb/metadata_schemas.rs
// metadata_schemas: Workspace-defined custom metadata field definitions.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum MetadataFieldType {
    Text,          // type: String
    Number,
    Date,
    Boolean,
    SingleSelect,
    MultiSelect,
}

// Composite unique(workspace_id, field_name) — enforce in reducer logic

#[spacetimedb::table(name = metadata_schemas, public)]
pub struct MetadataSchema {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub workspace_id: String, // FK → workspaces.id (cascade delete)

    pub field_name: String,
    pub field_label: String,
    pub field_type: MetadataFieldType,
    pub options: Option<String>, // JSON stored as string
    pub is_required: bool, // default false
    pub display_order: i32, // default 0

    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
